### PR TITLE
Configure nvhpc GCC from Spack variables

### PR DESCRIPTION
**Pull request by msimberg**
This is https://github.com/spack/spack/pull/24654 rebased on develop with a small modification to the patch. The reason for the changes is unchanged from https://github.com/spack/spack/pull/24654.

I've replaced the specific patch (https://github.com/spack/spack/pull/24654/files#diff-cf482bcc3e73486f18429a2cf2092deee184fe53c6d55cc464e325ebb044e838) with a patch function that is meant to apply to all versions and architectures. It works for me on x86_64. However, I have not tested this on aarch64 and ppc64le.

### CODE REVIEW

File: var/spack/repos/builtin/packages/nvhpc/package.py
Comment (samcmill) [inline var/spack/repos/builtin/packages/nvhpc/package.py]: With the call to makelocalrc with -gcc on line 152, there is no need to patch makelocalrc.

Comment (msimberg) [reply var/spack/repos/builtin/packages/nvhpc/package.py]: It is required with compilers that are defined only with the binary name, not the full path, like Cray's compilers in modules. See https://github.com/spack/spack/pull/24654#issue-934984728 and https://github.com/spack/spack/pull/24654#discussion_r814754290.

Comment (samcmill) [reply var/spack/repos/builtin/packages/nvhpc/package.py]: The danger of this is that it will pick and save whatever random compiler happens to be in your path at the time you installed the nvhpc package.

Comment (msimberg) [reply var/spack/repos/builtin/packages/nvhpc/package.py]: It's spack's job to ensure that the correct compiler is in `PATH` at the time of installing `nvhpc`. Without the patch it's guaranteed broken with module-based compilers. With the patch there's at least a chance of using the compiler specified by spack. For compilers that use the full path this doesn't change anything. I'd even argue that the whole line of adding `/usr/bin` and `/bin` to `PATH` could be removed for spack, since it's spack's job to ensure that everything needed is in `PATH`. If it's not it's either a bug in spack or someone has seriously broken their environment.

Comment (samcmill) [reply var/spack/repos/builtin/packages/nvhpc/package.py]: I may not be understanding https://github.com/spack/spack/pull/24654#discussion_r814754290.  It sounds like the environment module is setting `CC=cc`.  `makelocalrc` uses gcc, so I think it would ignore it and use the system gcc in that case, even with the patch?

For compilers defined in environment modules, I think you'd want to do `spack compiler find` to add that compiler to Spack, and then install the `nvhpc` package setup to use that Spack recognized compiler.  That will be more reproducible and predictable than just relying on whatever module happened to be loaded when you installed nvhpc.

Comment (msimberg) [reply var/spack/repos/builtin/packages/nvhpc/package.py]: > I may not be understanding [#24654 (comment)](https://github.com/spack/spack/pull/24654#discussion_r814754290). It sounds like the environment module is setting `CC=cc`.

Yes and no. Spack is setting the compiler to `CC`. This is `compilers.yaml` for one of the Cray GCCs:
```
compilers:
- compiler:
    spec: gcc@9.3.0
    paths:
      cc: cc
      cxx: CC
      f77: ftn
      fc: ftn
    flags: {}
    operating_system: cnl7
    target: any
    modules:
    - PrgEnv-gnu
    - gcc/9.3.0
    environment: {}
    extra_rpaths: []
```
As much as I'd like not to have compiler wrappers on Cray it's unfortunately still a reality. I'm not loading compiler modules manually and hoping that `spack install nvhpc` will use the correct compiler. I'm installing (e.g.) `nvhpc%gcc@9.3.0` and hoping that nvhpc will be configured with the correct GCC. Getting that to work requires:
1. passing `-gcc` and friends to `makelocalrc` to make it use spack's compiler wrapper, and
2. not prepending `/bin` and `/usr/bin` because doing that ends up with spack's compiler wrapper, which calls `CC`, which calls `g++`, to use the wrong `g++` if there's a `g++` in `/bin` or `/usr/bin`.

Does that make more sense?

Comment (samcmill) [reply var/spack/repos/builtin/packages/nvhpc/package.py]: It's number 2 that I don't understand.  Why is that necessary if `-gcc`, etc. are getting passed?

If you pass `-gcc`, etc. to `makelocalrc` then the PATH does not matter.  It just uses the values you provided, ignoring whatever is in the path.  The resulting `localrc` file does not save the path to `gcc`, or whatever compiler was used`, only the resulting include, lib, etc. paths (e.g., `<compliler> -print-search-dirs`).  

I confirmed the following works as expected without the PATH changes:
```
$ module load gcc/11.2.0
$ spack compiler find
$ spack install nvhpc%gcc@11.2.0
```

Comment (msimberg) [reply var/spack/repos/builtin/packages/nvhpc/package.py]: Does your compiler config end up having full paths to `g++` etc.?

> It's number 2 that I don't understand. Why is that necessary if `-gcc`, etc. are getting passed?

Because `self.compiler.cxx` is `CC` so it becomes `-gcc CC` in this case, not the full path to `g++`.

As mentioned earlier I dislike `CC`, but I also think it's a bad idea to prepend things to `PATH`. I think we'll be able to not use the compiler wrappers in the future on Cray systems but at the moment I think this is quite a common (spack) configuration on Cray systems.

Edit: the failure mode for this is that `nvhpc` uses the headers from the system GCC, which may be old. I noticed this while trying to use `nvc++ -std=c++17` and failing to compile things that clearly should be possible to compile with C++17, until I realized `nvhpc` had been configured with the system GCC (7) instead of the GCC in the spec.

Comment (samcmill) [reply var/spack/repos/builtin/packages/nvhpc/package.py]: If I'm understanding correctly, the `cc` wrapper relies on `PATH` to find gcc?  Something like 
```
export PATH=/path/to/other/gnu-compilers;$PATH
gcc $*
```
rather than
```
/path/to/other/gnu-compilers/gcc $*
```
?

If that's the general scheme, then I'm still confused why overriding `PATH` in the `cc` wrapper doesn't have precedence over reseting the path in `makelocalrc`?  `makelocalrc` is invoking `cc` (via `-gcc cc`), so `cc` can override it.

Is it something like `-gcc cc` ends up invoking `/usr/bin/cc` rather than `/path/to/cray-compilers/cc`?

Comment (msimberg) [reply var/spack/repos/builtin/packages/nvhpc/package.py]: Something like that. `CC` and `cc` are opaque binaries so I don't know exactly what goes on in there. I went back and checked, and it's specifically `cc` that creates problems, not `CC` (i.e. the C compiler, not the C++ compiler). On this system:

```
> module switch PrgEnv-cray PrgEnv-gnu
> module list |& grep gcc # GCC 11.2.0 is loaded
  3) gcc/11.2.0
> cc -print-search-dirs # all good
install: /opt/gcc/11.2.0/snos/lib/gcc/x86_64-suse-linux/11.2.0/
programs: =/opt/gcc/11.2.0/snos/libexec/gcc/x86_64-suse-linux/11.2.0/:/opt/gcc/11.2.0/snos/libexec/gcc/x86_64-suse-linux/11.2.0/:/opt/gcc/11.2.0/snos/libexec/gcc/x86_64-suse-linux/:/opt/gcc/11.2.0/snos/lib/gcc/x86_64-suse-linux/11.2.0/:/opt/gcc/11.2.0/snos/lib/gcc/x86_64-suse-linux/:/opt/gcc/11.2.0/snos/lib/gcc/x86_64-suse-linux/11.2.0/../../../../x86_64-suse-linux/bin/x86_64-suse-linux/11.2.0/:/opt/gcc/11.2.0/snos/lib/gcc/x86_64-suse-linux/11.2.0/../../../../x86_64-suse-linux/bin/
libraries: =/opt/gcc/11.2.0/snos/lib/gcc/x86_64-suse-linux/11.2.0/:/opt/gcc/11.2.0/snos/lib/gcc/x86_64-suse-linux/11.2.0/../../../../x86_64-suse-linux/lib/x86_64-suse-linux/11.2.0/:/opt/gcc/11.2.0/snos/lib/gcc/x86_64-suse-linux/11.2.0/../../../../x86_64-suse-linux/lib/../lib64/:/opt/gcc/11.2.0/snos/lib/gcc/x86_64-suse-linux/11.2.0/../../../x86_64-suse-linux/11.2.0/:/opt/gcc/11.2.0/snos/lib/gcc/x86_64-suse-linux/11.2.0/../../../../lib64/:/lib/x86_64-suse-linux/11.2.0/:/lib/../lib64/:/usr/lib/x86_64-suse-linux/11.2.0/:/usr/lib/../lib64/:/opt/gcc/11.2.0/snos/lib/gcc/x86_64-suse-linux/11.2.0/../../../../x86_64-suse-linux/lib/:/opt/gcc/11.2.0/snos/lib/gcc/x86_64-suse-linux/11.2.0/../../../:/lib/:/usr/lib/
> PATH=/usr/bin:/bin:$PATH cc -print-search-dirs # uh-oh
install: /usr/lib64/gcc/x86_64-suse-linux/7/ # this ends up as GCCDIR in makelocalrc
programs: =/usr/lib64/gcc/x86_64-suse-linux/7/:/usr/lib64/gcc/x86_64-suse-linux/7/:/usr/lib64/gcc/x86_64-suse-linux/:/usr/lib64/gcc/x86_64-suse-linux/7/:/usr/lib64/gcc/x86_64-suse-linux/:/usr/lib64/gcc/x86_64-suse-linux/7/../../../../x86_64-suse-linux/bin/x86_64-suse-linux/7/:/usr/lib64/gcc/x86_64-suse-linux/7/../../../../x86_64-suse-linux/bin/
libraries: =/usr/lib64/gcc/x86_64-suse-linux/7/:/usr/lib64/gcc/x86_64-suse-linux/7/../../../../x86_64-suse-linux/lib/x86_64-suse-linux/7/:/usr/lib64/gcc/x86_64-suse-linux/7/../../../../x86_64-suse-linux/lib/../lib64/:/usr/lib64/gcc/x86_64-suse-linux/7/../../../x86_64-suse-linux/7/:/usr/lib64/gcc/x86_64-suse-linux/7/../../../../lib64/:/lib/x86_64-suse-linux/7/:/lib/../lib64/:/usr/lib/x86_64-suse-linux/7/:/usr/lib/../lib64/:/usr/lib64/gcc/x86_64-suse-linux/7/../../../../x86_64-suse-linux/lib/:/usr/lib64/gcc/x86_64-suse-linux/7/../../../:/lib/:/usr/lib/
```

Edit: a bit more explicit
```
> PATH=/usr/bin:/bin:$PATH cc --version
cc (SUSE Linux) 7.5.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

> PATH=/usr/bin:/bin:$PATH CC --version
g++ (GCC) 11.2.0 20210728 (Cray Inc.)
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

Edit 2: and the module
```
> module show gcc
-------------------------------------------------------------------
/opt/modulefiles/gcc/11.2.0:

conflict         gcc
conflict         gcc-cross-aarch64
prepend-path     PATH /opt/gcc/11.2.0/bin
prepend-path     MANPATH /opt/gcc/11.2.0/snos/share/man
prepend-path     INFOPATH /opt/gcc/11.2.0/snos/share/info
prepend-path     LD_LIBRARY_PATH /opt/gcc/11.2.0/snos/lib64
setenv           GCC_PATH /opt/gcc/11.2.0
setenv           GCC_PREFIX /opt/gcc/11.2.0
setenv           GCC_VERSION 11.2.0
setenv           GNU_VERSION 11.2.0
-------------------------------------------------------------------
```
I suppose `cc` ends up calling `gcc`, whereas `CC` for some reason is not confused by the added `g++` in the path.

Comment (samcmill) [reply var/spack/repos/builtin/packages/nvhpc/package.py]: What does `which cc` return in those 2 cases?  I suspect that in the `PATH=/usr/bin:/bin/$PATH` case, `cc` is actually `/usr/bin/cc`, not `/opt/gcc/11.2.0/snos/bin/cc` (or whatever).  

Spack has a `which()` function.  What about something like this?
```
if self.compiler.cc is not an absolute path:
  gcc = which(self.compiler.cc)
  ...
else:
  gcc = self.compiler.cc

# pass gcc, etc. to makelocalrc
```

Comment (msimberg) [reply var/spack/repos/builtin/packages/nvhpc/package.py]: That's an excellent point, I did not think about that. That's exactly the case here.

However... I thought `self.compiler.cc` was simply `cc` defined in `compilers.yaml` but it's probably spack's compiler wrapper. `which(self.compiler.cc)` evaluated in `package.py` becomes `$SPACK_ROOT/lib/spack/env/cc`. I don't know if it's possible to get the absolute path of Cray's `cc` evaluated in some appropriate environment. I don't know what to think about this anymore.

Some input from @alalazo, @haampie, or some other spack core developer would be appreciated I think.

Finally, I _still_ don't understand what case `# Just in case, add /usr/bin and /bin to PATH` is trying to protect from.

Comment (samcmill) [reply var/spack/repos/builtin/packages/nvhpc/package.py]: > Finally, I _still_ don't understand what case `# Just in case, add /usr/bin and /bin to PATH` is trying to protect from.

`/usr/bin` and `/bin` are added to ensure that the various tools used in the script (`cat`, `sed`, etc.) are available.

They are prepended to `$PATH` so that the configuration is deterministic and reproducible.  Otherwise the install would be configured with whatever `gcc` happens to found in the user's environment, which may differ from another user's environment.  It may also differ from what the user intended to use, not understanding that having a gcc module loaded will impact the nvhpc install configuration.

Using alternative GNU compilers is an opt in by explicitly using the `-gcc`, etc. options.

So the change to add support for the `-gcc`, etc. options is good from my perspective.  It will make `spack install nvhpc%gcc@X.Y.Z` work as expected.  

On the other hand, the patch to remove the prepends to the `$PATH` is problematic.  I understand the case now, but I think the real bug that needs to be addressed is that the Cray compilers should be configured with an absolute path.

I advise separating this into 2 PRs, one with for the `-gcc` option that can be merged now, and another for the `$PATH` changes that will require further discussion.

Comment (haampie) [reply var/spack/repos/builtin/packages/nvhpc/package.py]: Reading this discussion:

Modifying `PATH` is a mistake in the context of Spack, as Mikael explained above. Even outside of Spack, it's extremely unlikely there are non-posix unix tools in the PATH, and much more likely that whenever there's a "non-standard" `gcc` in the PATH, that is intentional. Now for Spack, the PATH is setup correctly, don't change it. And regarding Cray systems, I suppose it's using `cc` not `/path/to/cray/cc` so that  Spack configurations survive updates to the programming environment.

I don't see the need to split this PR up into separate ones, both changes make sense in the context of Spack.

Comment (samcmill) [reply var/spack/repos/builtin/packages/nvhpc/package.py]: > Even outside of Spack, it's extremely unlikely there are non-posix unix tools in the PATH, and much more likely that whenever there's a "non-standard" `gcc` in the PATH, that is intentional. 

The `$PATH` modification is there because have been systems / environments where this has been a problem.

And it is not true that having a non-standard gcc in the PATH is always intentional.  Many will not realize the standard library headers are used from the GNU installation.  Prepending the system PATH makes the behavior reproducible, rather than caching whatever GNU environment happened to be active when nvhpc was installed.

Granted Spack has its own methods of controlling for this.  But apparently not 100% since the `cc` used by the Cray compilers can be modified outside of Spack.

> I don't see the need to split this PR up into separate ones, both changes make sense in the context of Spack.

I suggested this because I'm ready to sign off on the `-gcc` change, but not the `$PATH` change.

Comment (msimberg) [reply var/spack/repos/builtin/packages/nvhpc/package.py]: @samcmill I've removed the `PATH` patch from this PR. We can handle that separately. Do the remaining changes look ok to you?

Comment (samcmill) [reply var/spack/repos/builtin/packages/nvhpc/package.py]: LGTM

Comment (haampie) [reply var/spack/repos/builtin/packages/nvhpc/package.py]: I'm about to force push some changes `os.system` -> `Executable`, `os.environ` -> `setup_build_environment`, and adding back the `-net` flag to `makerc`.

**Comment by haampie** [comment]
Okay, so now spack also prints the `./install` and `makelocalrc args...` during the install step, and properly sets env variables.

**Comment by msimberg** [comment]
I can't approve this since it's my PR, but I approve of @haampie's changes...

**Comment by samcmill** [comment]
LGTM

**Review by alalazo — APPROVED** [review]
Approved by @samcmill Package modifications LGTM

**Pull request merged**
